gaia: init at 2.4.5

doronbehar opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

doronbehar committed:
gaia: init at 2.4.5

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+, qt4
+, python
+, pythonSupport ? true
+# Default to false since it brakes the build
```
```suggestion
# Default to false since it breaks the build
```

What's the reason for the build failure?

doronbehar replied:
See https://github.com/MTG/gaia/issues/11, I've added this link in the comment.

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+  '';
+
+  # This is not exactly specified in upstream's README but it's needed by the
+  # resultings $out/bin/gaiafusion script
```
```suggestion
  # resulting $out/bin/gaiafusion script
```

doronbehar replied:
Thanks :)

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+  postInstall = (if pythonSupport then ''
+    # We can't use patchShebangs because it will use bare bones $python/bin/python
+    # and we need a python environment with pyyaml
+    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${python.sitePackages}:${pythonEnv}/${python.sitePackages}
```
Why not use `wrapPython` instead? Using an extra python environment seems overkill if you just need `pyyaml`.

doronbehar replied:
Sounds like a good idea, I didn't know this existed. I wouldn't mind giving up upon the `pythonEnv` and use the dependencies straight in the inputs. However, whether I use a `pythonEnv` or not, the harder part is that the `$out/bin/gaiafusion` script requires the `$PYTHONPATH` to include gaia's python modules - those in `$out/${python.sitePackages}`.

I tried the following:

```patch
diff --git i/pkgs/development/libraries/gaia/default.nix w/pkgs/development/libraries/gaia/default.nix
index dcf44ec6ca4..5a060f5001b 100644
--- i/pkgs/development/libraries/gaia/default.nix
+++ w/pkgs/development/libraries/gaia/default.nix
@@ -36,26 +36,25 @@ stdenv.mkDerivation rec {
       --replace "/etc/init.d" "$out/etc/init.d"
   '';
 
-  # This is not exactly specified in upstream's README but it's needed by the
-  # resultings $out/bin/gaiafusion script
-  pythonEnv = (if pythonSupport then
-    python.withPackages(ps: with ps; [
-      pyyaml
-    ])
-  else null);
-
   nativeBuildInputs = [
     wafHook
     pkgconfig
     swig
     makeWrapper
-  ];
+  ]
+    # The gaiafusion binary inside $out/bin needs a shebangs patch, and
+    # wrapping with the appropriate $PYTHONPATH
+    ++ lib.optionals (pythonSupport) [ pythonPackages.wrapPython ]
+  ;
   buildInputs = [
     libyaml
     qt4
   ]
     ++ lib.optionals (pythonSupport) [
-      pythonEnv
+      pythonPackages.python
+      # This is not exactly specified in upstream's README but it's needed by the
+      # resulting $out/bin/gaiafusion script
+      pythonPackages.pyyaml
     ]
   ;
   wafConfigureFlags = [
@@ -65,12 +64,6 @@ stdenv.mkDerivation rec {
     ++ lib.optionals (assertsSupport) [ "--with-asserts" ]
     ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
   ;
-  # only gaiafusion is a python executable that needs patchShebangs
-  postInstall = lib.optionalString ''
-    # We can't use patchShebangs because it will use bare bones $python/bin/python
-    # and we need a python environment with pyyaml
-    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${python.sitePackages}:${pythonEnv}/${python.sitePackages}
-  '';
 
   meta = with lib; {
     homepage = "https://github.com/MTG/gaia";
```

But `$out/bin/gaiafusion` comes out as if I had used a bare `patchShebangs` - it doesn't find gaia's modules.

I tried also using a `postInstall` including `wrapPythonPrograms` but then `$out/bin/gaiafusion` couldn't find `yaml`.

I tried also using `wrapPythonProgramsIn $out/bin/ $PYTHONPATH:$out/${pythonPackages.python.sitePackages}` but then `$out/bin/gaiafusion` couldn't find gaia's modules as well.

I noticed that the shell function `wrapPythonProgramsIn` doesn't send `$PYTHONPATH` to `wrapProgram`. See:

https://github.com/NixOS/nixpkgs/blob/a3f24daa7bc52767e15b5476d6c6b4165db5ee39/pkgs/development/interpreters/python/wrap.sh#L71-L77

I'd expect that shell array to include:

```
--prefix PYTHONPATH : "$program_PYTHONPATH"
```

I tried adding these strings to the array but that turned out to trigger a massive rebuild of almost every package so I gave up. I did try to use `makeWrapperArgs` in the derivation:

```nix
  makeWrapperArgs = [
    "--prefix" "PYTHONPATH" ":" "$program_PYTHONPATH"
  ];
```

Hoping `wrapPythonProgramsIn` will consider it when writing `wrapProgram`. But then `$out/bin/gaiafusion` didn't find gaia's modules as well :disappointed:.

It's kind of disappointing none of these shell functions work and that I have to use the good old `wrapProgram` my self. It's not that bad TBH, but it's not as elegant as well.

aszlig replied:
Ah, I should have been more clear on this, here is a diff of it working:

```diff
diff --git a/pkgs/development/libraries/gaia/default.nix b/pkgs/development/libraries/gaia/default.nix
index 7565ee7a06e..9367792e06c 100644
--- a/pkgs/development/libraries/gaia/default.nix
+++ b/pkgs/development/libraries/gaia/default.nix
@@ -36,41 +36,23 @@ stdenv.mkDerivation rec {
       --replace "/etc/init.d" "$out/etc/init.d"
   '';
 
-  # This is not exactly specified in upstream's README but it's needed by the
-  # resulting $out/bin/gaiafusion script
-  pythonEnv = (if pythonSupport then
-    pythonPackages.python.withPackages(ps: with ps; [
-      pyyaml
-    ])
-  else null);
-
   nativeBuildInputs = [
     wafHook
     pkgconfig
     swig
     makeWrapper
-  ];
-  buildInputs = [
-    libyaml
-    qt4
-  ]
-    ++ lib.optionals (pythonSupport) [
-      pythonEnv
-    ]
-  ;
-  wafConfigureFlags = [
-  ]
-    ++ lib.optionals (pythonSupport) [ "--with-python-bindings" ]
-    ++ lib.optionals (stlfacadeSupport) [ "--with-stlfacade" ]
-    ++ lib.optionals (assertsSupport) [ "--with-asserts" ]
-    ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
-  ;
-  # only gaiafusion is a python executable that needs patchShebangs
-  postInstall = lib.optionalString (pythonSupport) ''
-    # We can't use patchShebangs because it will use bare bones $python/bin/python
-    # and we need a python environment with pyyaml
-    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${pythonPackages.python.sitePackages}:${pythonEnv}/${pythonPackages.python.sitePackages}
-  '';
+  ] ++ lib.optional pythonSupport pythonPackages.wrapPython;
+
+  buildInputs = [ libyaml qt4 ];
+
+  propagatedBuildInputs = lib.optional pythonSupport pythonPackages.pyyaml;
+
+  wafConfigureFlags = lib.optional pythonSupport "--with-python-bindings"
+                   ++ lib.optional stlfacadeSupport "--with-stlfacade"
+                   ++ lib.optional assertsSupport "--with-asserts"
+                   ++ lib.optional cyclopsSupport "--with-cyclops";
+
+  postFixup = "wrapPythonPrograms";
 
   meta = with lib; {
     homepage = "https://github.com/MTG/gaia";
```

The point here is that wrapping Python programs is only done automatically if you actually use the `buildPython*` functions. What's different in the above is the additional `postFixup` hook that's calling `wrapPythonPrograms`, which then will add all Python inputs in `propagatedBuildInputs` to the corresponding executables.

doronbehar replied:
I see. Thanks for the patch! I've applied a version of mine to the branch and pushed the changes. The only differences are that I used `lib.optionals` and other styles of writing we are do not agree on :smile:.

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+      pythonEnv
+    ]
+  ;
+  wafConfigureFlags = [
```
Superfluous empty list.

doronbehar replied:
Same as [here](https://github.com/NixOS/nixpkgs/pull/70372/files/e86ef91b58d23d28c327b59c9ed4b6041bd8c9cb#r331733095).

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+  };
+
+  # Fix installation error when waf tries to put files in /etc/
+  prePatch = ''
```
Superfluous empty string.

doronbehar replied:
Adding all of the superfluous empty strings/lists is kind of a habit of mine. The idea is, that if at some point in the future we'll add e.g a `prePatch` that will always be set, the diff is going to be much clearer.

aszlig replied:
Well, I beg to differ, because while it avoids one line in the diff, it makes the code harder to read and I tend to prefer avoiding the latter, but anyway, it's just a matter of personal preference and as long we don't have a code style guide that explicitly bans this I'm fine.

doronbehar replied:
> as long we don't have a code style guide that explicitly bans this I'm fine.

I wonder whether `nix-fmt` will handle this when the day comes and it'll work in production..

I hope it's OK with you that I'll mark this as resolved although no change was done. I wouldn't mind that much changing it if another reviewer will say he doesn't like it.

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+  ;
+  wafConfigureFlags = [
+  ]
+    ++ lib.optionals (pythonSupport) [ "--with-python-bindings" ]
```
As mentioned in https://github.com/NixOS/nixpkgs/pull/70389/files#r331696607, use of `lib.optional` here will already imply a singleton list.

doronbehar replied:
I've had comments in the past saying that `lib.optionals` should always be preferred over `lib.optional`. See https://github.com/NixOS/nixpkgs/pull/65480#discussion_r315800912

aszlig replied:
Again, similar to the superfluous empty lists/strings it's just personal opinion, but I do not find `lib.optional` a bad thing, because the less scanning one needs to do for `[` and `]` the better IMO. But again, it's a matter of personal preference, and I also tend to use `lib.singleton` a lot, where other people would probably shriek in horror ;-)

doronbehar replied:
I see, my main motive is that future diffs will be clean and I'd like at least to be consistent with my self.. I appreciate your opinion though. Allow me to mark this conversation as resolved based upon our agreement upon disagreement :relaxed: .

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+    ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
+  ;
+  # only gaiafusion is a python executable that needs patchShebangs
+  postInstall = (if pythonSupport then ''
```
```suggestion
  postInstall = lib.optionalString pythonSupport ''
```

doronbehar replied:
:+:

aszlig commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,81 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/MTG/gaia";
+    description = "C++ library with python bindings which implements similarity measures and classiﬁcations on the results of audio analysis, and generates classiﬁcation models that Essentia can use to compute high-level description of music";
```
This should only be a a [short description](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes), so I'd suggest something like this (from the upstream [debian/control](https://github.com/MTG/gaia/blob/95f4851d051ed5ef59d668796fb92319393a63dc/debian/control#L11) file):

```suggestion
    description = "General library to work with points in a semimetric space";
```

doronbehar replied:
Yea I wasn't sure about that long description. Good idea to use upstream's description :+:.

doronbehar committed:
Link upstream issue about stlfacadeSupport defaulting false

doronbehar committed:
Use upstream's short description in meta

doronbehar committed:
Use lib.optionalString for postInstall

doronbehar committed:
Improve pythonSupport handling

Use pythonPackages instead of just python and add an attribute to
python-packages.nix.

doronbehar reviewed (commented):
Thanks for all the comments @aszlig, I've commented as well and pushed changes.

doronbehar committed:
Add platforms to meta

doronbehar committed:
Make gaia unsupported with python2

doronbehar committed:
Use wrapPython in postFixup instead of wrapProgram

doronbehar reviewed (commented):
Thanks @aszlig for the comments, I've pushed a working build including a disable for Python3.

doronbehar commented on pkgs/development/libraries/gaia/default.nix:
```diff
@@ -0,0 +1,85 @@
+    substituteInPlace src/wscript \
+      --replace "/etc/cyclops" "$out/etc/cyclops" \
+      --replace "/etc/init.d" "$out/etc/init.d"
+  '';
```
BTW @aszlig, what do you think about the fact the package installs `init.d` files? I've had a look at these files today and I noticed they refer to paths which rely upon the fact the system will comply to the FHS hierarchy. Patching them might require some work. I wouldn't mind dive into it but NixOS doesn't even use these init files so what would be the point of that? Should we provide systemd files that will provide the same functionality? TBH I don't even know what these daemons are for and I've never used them not I intend to. Maybe we can keep it as is so future users of this package will be able to look at them for reference?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/71

doronbehar commented:
Thanks @matthewbauer! Now that this is merged, could you please take a look at #70389?
